### PR TITLE
Misc build fixes

### PR DESCRIPTION
**Pull request by ellismg**
This is a set of infrastructure changes to make it easier to consume Roslyn as part of the overall .NET Core build from source effort.

These commits make `cibuild.sh` suitable for building the relevant parts of Roslyn for .NET Core on all Linux distributions we currently support.  I've also added the ability to produce nuget packages on *NIX instead of just windows.

/cc @jaredpar

### CODE REVIEW

File: CrossPlatform.sln [hunk 1]
```diff
@@ -130,8 +138,8 @@ Global
 		{1EE8CAD3-55F9-4D91-96B2-084641DA9A6C}.Debug|ARM.Build.0 = Debug|Any CPU
 		{1EE8CAD3-55F9-4D91-96B2-084641DA9A6C}.Debug|Mixed Platforms.ActiveCfg = Debug|Any CPU
 		{1EE8CAD3-55F9-4D91-96B2-084641DA9A6C}.Debug|Mixed Platforms.Build.0 = Debug|Any CPU
-		{1EE8CAD3-55F9-4D91-96B2-084641DA9A6C}.Debug|x64.ActiveCfg = Debug|x64
-		{1EE8CAD3-55F9-4D91-96B2-084641DA9A6C}.Debug|x64.Build.0 = Debug|x64
+		{1EE8CAD3-55F9-4D91-96B2-084641DA9A6C}.Debug|x64.ActiveCfg = Debug|Any CPU
```
Comment (ellismg) [inline CrossPlatform.sln]: These diffs came from just adding the projects using VS2015. I'm not sure if moving from x64 to Any CPU is a problem.  If you want, I can manually revert these hunks.

File: cibuild.sh [hunk 2]
```diff
@@ -7,14 +7,19 @@ usage()
...
     echo "Options"
-    echo "  --debug     Build Debug (default)"
-    echo "  --release   Build Release"
-    echo "  --nocache   Force download of toolsets"
-
+    echo "  --debug               Build Debug (default)"
+    echo "  --release             Build Release"
+    echo "  --skiptests           Do not run tests"
+    echo "  --skipcrossgen        Do not crossgen the bootstrapped compiler"
+    echo "  --skipcommitprinting  Do not print commit information"
```
Comment (ellismg) [inline cibuild.sh]: If you would rather me just move the commit printing out of here and into the CI definition, I'm happy to do that.  I didn't really understand the point of doing it here (shouldn't the Jenkins logs already provide the information of the SHA you are building?)

This causes problems both on RHEL (where there git is older and doesn't understand `--no-patch` and will be problematic when the sources are not in a git repository).

File: src/NuGet/BuildNuGets.csx [hunk 3]
```diff
@@ -1,4 +1,4 @@
-#r "System.Xml.Linq"
+#r "System.Xml.XDocument.dll"
```
Comment (ellismg) [inline src/NuGet/BuildNuGets.csx]: This had me a little worried.  What's the best way for me to ensure this isn't going to break how this is used on Windows to produce nupkgs?  Do I get validation from CI?

File: src/NuGet/BuildNuGets.csx [hunk 4]
```diff
@@ -126,6 +129,24 @@ var PreReleaseOnlyPackages = new HashSet<string>
     "Microsoft.VisualStudio.LanguageServices.Next",
 };
 
+// The assets for these packages are not produced when building on Unix
+// and we don't want to attempt to package them when building packages.
+var PackagesNotBuiltOnCore = new HashSet<string>
+{
+     "Microsoft.CodeAnalysis.CSharp.Features",
```
Comment (ellismg) [inline src/NuGet/BuildNuGets.csx]: I generated this list by just including what failed due to missing build outputs.  You can tell me if it looks sensible or not.

File: src/NuGet/Microsoft.CodeAnalysis.Build.Tasks.nuspec [hunk 5]
```diff
@@ -46,9 +46,9 @@
     </contentFiles>
   </metadata>
   <files>
-    <file src="Dlls\MSBuildTask\Microsoft.Build.Tasks.CodeAnalysis.dll" target="lib\netstandard1.3" />
-    <file src="Dlls\MSBuildTask\Microsoft.CSharp.Core.targets" target="contentFiles\any\any" />
-    <file src="Dlls\MSBuildTask\Microsoft.VisualBasic.Core.targets" target="contentFiles\any\any" />
+    <file src="Dlls/MSBuildTask/Microsoft.Build.Tasks.CodeAnalysis.dll" target="lib/netstandard1.3" />
```
Comment (ellismg) [inline src/NuGet/Microsoft.CodeAnalysis.Build.Tasks.nuspec]: I think there are some other NuSpecs in the repository just outside of this folder that I can also convert if that's helpful.

**Comment by ellismg** [comment]
Looks like I forgot to run RepoUtils's change command (it can't be run on *NIX today) so I'll do that.  @jaredpar is there additional testing I should be doing on windows around the small changes I made to the NuGet package logic?

**Comment by jaredpar** [comment]
CC @agocke, @TyOverby, @OmarTawfik

File: CrossPlatform.sln [hunk 6]
```diff
@@ -91,6 +91,14 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "MSBuildTask", "src\Compiler
...
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "ScriptingTestUtilities", "src\Scripting\CoreTestUtilities\ScriptingTestUtilities.csproj", "{21A01C2D-2501-4619-8144-48977DD22D9C}"
 EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Interactive", "Interactive", "{FED79E76-76A7-48A1-B0F1-E5E56B5E7FE4}"
+EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Hosts", "Hosts", "{0F0F4F95-0BEC-4623-9B26-DBD662F1193F}"
+EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "CsiCore", "src\Interactive\CsiCore\CsiCore.csproj", "{D1B051A4-F2A1-4E97-9747-C41D13E475FD}"
+EndProject
+Project("{F184B08F-C81C-45F6-A57F-5ABD9991F28F}") = "VbiCore", "src\Interactive\VbiCore\VbiCore.vbproj", "{1EEFB4B6-A6CC-4869-AF05-A43C8B82A8FD}"
+EndProject
```
Comment (jaredpar) [inline CrossPlatform.sln]: A bit surprised that we need that.  What does it get included in?

File: CrossPlatform.sln (hunk 1 above)
Comment (jaredpar) [inline CrossPlatform.sln]: These entries are a mystery to me on a good day.

File: cibuild.sh (hunk 2 above)
Comment (jaredpar) [inline cibuild.sh]: I can't remember the reason we did the commit printing in the first place.  @agocke may remember.

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 7]
```diff
@@ -4191,9 +4191,6 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   <data name="IDS_FeatureRefLocalsReturns" xml:space="preserve">
     <value>byref locals and returns</value>
   </data>
-  <data name="IDS_ThrowExpression" xml:space="preserve">
-    <value>&lt;throw&gt;</value>
-  </data>
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: Still can't find a good reason that the warning from this duplication doesn't show up on Windows.  Looked around and saw other reports of it. 

Anyways, thanks for removing it.

File: src/NuGet/BuildNuGets.csx (hunk 3 above)
Comment (jaredpar) [inline src/NuGet/BuildNuGets.csx]: The CI job of microbuild validates our NuGet package building.

File: src/NuGet/Microsoft.CodeAnalysis.Build.Tasks.nuspec (hunk 5 above)
Comment (jaredpar) [inline src/NuGet/Microsoft.CodeAnalysis.Build.Tasks.nuspec]: Please convert any others as well that you know of.  Really I should just move them all in here though.

File: src/Tools/CommonNetCoreReferences/project.json [hunk 8]
```diff
@@ -34,6 +34,8 @@
         "System.Runtime.InteropServices": "4.3.0",
         "System.Runtime.Loader": "4.3.0",
         "System.Runtime.Numerics": "4.3.0",
+        "System.Runtime.Serialization.Primitives": "4.3.0",
+        "System.Runtime.Serialization.Json": "4.3.0",
```
Comment (jaredpar) [inline src/Tools/CommonNetCoreReferences/project.json]: Curious: why did these become necessary?

**Comment by jaredpar** [comment]
CC @tmat as well since he wrote the NuGet building file.

File: src/Tools/CommonNetCoreReferences/project.json (hunk 8 above)
Comment (ellismg) [inline src/Tools/CommonNetCoreReferences/project.json]: Without these using `csi` to run the BuildNuGets file was not working when hosted on core.

File: CrossPlatform.sln (hunk 6 above)
Comment (ellismg) [inline CrossPlatform.sln]: I want to use `csi` to run the NuGet building gunk.  I included `vbi` for symmetry.  I can remove it if you'd like, but I do need `csi`

Comment (jaredpar) [inline CrossPlatform.sln]: I just was curious if you actually needed it for something that I missed. 

I'd actually prefer to keep it.  One step closer to my goal of merging Compilers.sln and CrossPlatform.sln.

**Comment by ellismg** [comment]
@jaredpar Can you help me understand how to fix the RepoUtil errors in windows_eta_open_prtest?  It is complaining about a dependency I didn't touch and may have something to do with open vs closed code.  I'm a bit of a loss.  I think the xUnit issues in that leg are cascading failures from RepoUtil blowing up.

**Comment by jaredpar** [comment]
@ellismg ignore that for now.  Another changed was incorrectly merged which is causing that problem.

**Comment by ellismg** [comment]
The MicroBuild leg is spewing all sorts of warnings like this that makes me nervous:

```
16:13:04   EXEC : warning : Issue found with package 'Microsoft.CodeAnalysis.Build.Tasks'. [D:\j\workspace\microbuild_pr---d013602c\src\NuGet\NuGet.proj]
16:13:04   EXEC : warning : Issue: Assembly outside lib folder. [D:\j\workspace\microbuild_pr---d013602c\src\NuGet\NuGet.proj]
16:13:04   EXEC : warning : Description: The assembly 'lib/netstandard1.3\Microsoft.Build.Tasks.CodeAnalysis.dll' is not inside the 'lib' folder and hence it won't be added as reference when the package is installed into a project. [D:\j\workspace\microbuild_pr---d013602c\src\NuGet\NuGet.proj]
```

I'm going to dig into this more.  If I'm not sure this isn't going to break a bunch of stuff, I will probably move out that commit and update my wrapper build to just fix slash issues in NuSpec's locally before packing...

File: CrossPlatform.sln (hunk 6 above)
Comment (ellismg) [inline CrossPlatform.sln]: SGTM!

File: src/NuGet/BuildNuGets.csx [hunk 9]
```diff
@@ -145,32 +166,58 @@ void ReportError(string message)
 
 int PackFiles(string[] nuspecFiles, string licenseUrl)
 {
-    string commonArgs =
-        $"-BasePath \"{BinDir}\" " +
+    var commonProperties = new Dictionary<string, string>()
```
Comment (tmat) [inline src/NuGet/BuildNuGets.csx]: Why not just ```List<(string, string)>```?

Comment (ellismg) [inline src/NuGet/BuildNuGets.csx]: Have not internalized tuples yet.  I can change it to that.

File: src/NuGet/Microsoft.CodeAnalysis.Build.Tasks.nuspec (hunk 5 above)
Comment (tmat) [inline src/NuGet/Microsoft.CodeAnalysis.Build.Tasks.nuspec]: NuGet doesn't treat ```\``` and ```/``` the same like msbuild does? That's unfortunate.

File: src/Tools/CommonNetCoreReferences/project.json (hunk 8 above)
Comment (tmat) [inline src/Tools/CommonNetCoreReferences/project.json]: BTW, this https://github.com/dotnet/roslyn/pull/15658 is needed to run scripts on Core 1.1

**Comment by tmat** [comment]
👍

**Comment by jaredpar** [comment]
@ellismg looks like some conflicts crept into the master branch.  Once you resolve those we should be good to merge.

**Comment by ellismg** [comment]
@jaredpar I rebased to address the conflicts (formatting in project.json's) and I also dropped the nuspec path separator changes for now.

**Comment by jaredpar** [comment]
@ellismg looks like you changed some NuGet references.  Can you run the following:

``` 
Binaries\Debug\Exes\RepoUtil\RepoUtil change
```

That will re-generate the appropriate files and fix the two failing legs.

**Comment by jaredpar** [comment]
@ellismg looks like another conflict creeped in on BuildNuGets.csx.  You seem to be picking very popular files 😦

**Pull request merged**